packages/mql-typescript/src/generator.ts

```diff
@@ -0,0 +1,292 @@
+          }
+
+          throw new Error(
+            `Expected bson.BSONRegExp, but got ${data.constructor.name}`,
+          );
+        },
+      }),
+      new yaml.Type('!bson_binary', {
+        kind: 'scalar',
+        construct(data) {
+          return bson.Binary.createFromBase64(data);
+        },
+        predicate(data) {
+          return data instanceof bson.Binary;
+        },
+        represent(data) {
+          if (data instanceof bson.Binary) {
+            return data.toString('base64');
+          }
+
+          throw new Error(
+            `Expected bson.Binary, but got ${data.constructor.name}`,
+          );
+        },
+      }),
+      new yaml.Type('!bson_decimal128', {
+        kind: 'scalar',
+        construct(data) {
+          return bson.Decimal128.fromString(data);
+        },
+        predicate(data) {
+          return data instanceof bson.Decimal128;
+        },
+        represent(data) {
+          if (data instanceof bson.Decimal128) {
+            return data.toString();
+          }
+
+          throw new Error(
+            `Expected bson.Decimal128, but got ${data.constructor.name}`,
+          );
+        },
+      }),
+    ]),
+  };
+
+  protected configDir = path.join(
+    __dirname,
+    '..',
+    'mongo-php-library',
+    'generator',
+    'config',
+  );
+
+  private async *listCategories(
+    filterRegex: RegExp | undefined,
+  ): AsyncIterable<{
+    category: string;
+    folder: string;
+  }> {
+    for await (const folder of await fs.readdir(this.configDir, {
+      withFileTypes: true,
+    })) {
+      if (folder.isDirectory() && filterRegex?.test(folder.name) !== false) {
+        yield {
+          category: folder.name,
+          folder: path.join(folder.parentPath, folder.name),
+        };
+      }
+    }
+  }
+
+  private async *listSourceYAMLFiles(
+    categoryRegex: RegExp | undefined,
+    operatorRegex: RegExp | undefined,
+  ): AsyncIterable<{
+    category: string;
+    operators: () => AsyncIterable<{ yaml: unknown; path: string }>;
+  }> {
+    for await (const { category, folder } of this.listCategories(
+      categoryRegex,
+    )) {
+      yield {
+        category,
+        operators: async function* () {
+          for await (const file of await fs.readdir(folder, {
+            withFileTypes: true,
+          })) {
+            if (
+              file.isFile() &&
+              file.name.endsWith('.yaml') &&
+              operatorRegex?.test(file.name) !== false
+            ) {
+              const filePath = path.join(file.parentPath, file.name);
+              const content = await fs.readFile(filePath, 'utf8');
+              const parsed = yaml.load(content, GeneratorBase.loadOptions);
+
+              yield { yaml: parsed, path: filePath };
+            }
+          }
+        },
+      };
+    }
+  }
+
+  protected emitToFile(filePath: string): void {
+    this.outputStream = createWriteStream(filePath, { encoding: 'utf8' });
```

Comment:
Must we use streams? I find it kinda redundant in most cases nowadays with async/await, async iterators, etc. It makes it tricky to try and do multiple things with the data at once. For example I want to also write the types we generate as a giant string elsewhere and streams force all the file writing into the whole structure of this class, making it tricky to just do two things or do one thing after another without restructuring it all.